Quick_Deploy/vLLM/model_repository/vllm/1/model.py

```diff
@@ -0,0 +1,212 @@
+
+    def engine_loop(self, loop):
+        """
+        Runs the engine's event loop on a separate thread.
+        """
+        asyncio.set_event_loop(loop)
+        self._loop.run_until_complete(self.await_shutdown())
+
+    async def await_shutdown(self):
+        """
+        Primary coroutine running on the engine event loop. This coroutine is responsible for
+        keeping the engine alive until a shutdown is requested.
+        """
+        # first await the shutdown signal
+        while self._shutdown_event.is_set() is False:
+            await asyncio.sleep(5)
+
+        # Wait for the ongoing_requests
+        while self.ongoing_request_count > 0:
+            self.logger.log_info(
+                "Awaiting remaining {} requests".format(self.ongoing_request_count)
+            )
+            await asyncio.sleep(5)
+
+        self.logger.log_info("Shutdown complete")
+
+    def get_sampling_params_dict(self, params_json):
+        """
+        This functions parses the dictionary values into their
+        expected format.
+        """
+
+        dict = json.loads(params_json)
+
+        # Special parsing for the supported sampling parameters
+        # TODO: Add more parameters if needed
+        dict["temperature"] = float(dict["temperature"])
+        dict["top_p"] = float(dict["top_p"])
+
+        return dict
+
+    def create_response(self, vllm_output):
+        """
+        Parses the output from the vLLM engine into Triton
+        response.
+        """
+        prompt = vllm_output.prompt
+        text_outputs = [
+            (prompt + output.text).encode("utf-8") for output in vllm_output.outputs
+        ]
+        triton_output_tensor = pb_utils.Tensor(
+            "TEXT", np.asarray(text_outputs, dtype=self.output_dtype)
+        )
+        return pb_utils.InferenceResponse(output_tensors=[triton_output_tensor])
+
+    async def generate(self, request):
+        """
+        Forwards single request to LLM engine and returns responses.
+        """
+        response_sender = request.get_response_sender()
+        self.ongoing_request_count += 1
+        try:
+            request_id = random_uuid()
+            prompt = pb_utils.get_input_tensor_by_name(request, "PROMPT").as_numpy()[0]
+            stream = pb_utils.get_input_tensor_by_name(request, "STREAM").as_numpy()[0]
+            sampling_params_dict = self.get_sampling_params_dict(request.parameters())
+            sampling_params = SamplingParams(**sampling_params_dict)
+
+            last_output = None
+            async for output in self.llm_engine.generate(
+                str(prompt), sampling_params, request_id
+            ):
+                if stream:
+                    response_sender.send(self.create_response(output))
+                else:
+                    last_output = output
+
+            if not stream:
+                response_sender.send(self.create_response(last_output))
```

Comment:
If I'm understanding `not stream` correctly, this means we will send all generated tokens as a single response, right? If so, shouldn't we be concatenating the generated tokens instead of overwriting them in `last_output` ?

Reply:
The vLLM engine seems to be building the text output response with each output iteration. The last output seems to hold the entire response.
https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/api_server.py#L58-L71